docs/user/interop.rst

```diff
@@ -357,6 +357,9 @@ strings (similarly to C):
     val msg: CString = c"Hello, world!"
     stdio.printf(msg)
 
+It does not allow any octal values or escape characters not supported by Scala compiler, like ``\a`` or ``\?``.
+It is possible to use C-style hex values up to value 0xFF, eg. ``c"Hello \x61\x62\x63"``
+
```

Comment:
Well said!  I like the examples, they clarify.

I hope that you are not getting frustrated by the review cycles.

Whack a mole time!    I can create an independent issue for the concern below
so that the current PR can proceed. 

I think this section exposes a weakness in the section which follows it. 

Absent any documentation, it is reasonable to believe that the c interpolator
and toCString() have the same results.  One knows that _someone_ is going
to do toCString() and report that it allows escapes which the quote-equivalent-unquote
interpolator does not.
```
Additionally, we also expose two helper functions unsafe.toCString and unsafe.fromCString to convert between C-style and Java-style strings.
```
I think that creating some test cases to figure out what toCString() is actually
doing with regards to escapes and documenting it in this .rst is worthwhile.
I _think_ it is converting from Java strings to bytes and then copying those bytes to allocated memory.  The creation of the Java string would have done escape
processing but toCString() would not. Thus one would need a Java "\\a" to pass
 the two bytes '\' & 'a' to C.  

The same consideration and need for explicit test cases applies to fromCString().
For the sake of this discussion, consider a byte array read in from user input.
I believe that each byte gets visited one by one and converted to a Java 16 bit Character without escape processing.  Thus a CString holding the two bytes
'\' and 'n' become two Java Characters, not one newline.  

I propose that one or more test cases be created and the behavior documented
in this section of the .rst.

Reply:
The key insight is that:

* A `String` is a UCS string, i.e., a sequence of 2-byte Chars. It is *usually* interpreted as UTF-16.
* A `CString` is a *byte string*, i.e., a sequence of Bytes. It is *usually* interpreted as UTF-8, but it can also be ASCII or latin1 or whatever.
* The `c"..."` interpolator describes a *byte string*, without any specific encoding. Because the source code is text, *some* encoding needs to be chosen when encoding the text *at compile-time* into a byte string. The natural choice is latin1 aka ISO-8859-1, because it uses 1-byte code points, and its valid code point range (0-255) corresponds to the same range in UTF-16 `Char`s and in abstract Unicode code point.
* `fromCString` and `toCString` are charset-aware. They will always assume that the `String` is UTF-16 (as do the `java.nio.charset` APIs) and take an `Charset` parameter to know what encoding to assume for the byte string. If not specified, it is UTF-8. They are very similar to `String.getBytes` and `new String(Array[Byte])`, except that they work with a `CString` instead of an `Array[Byte]`.

`fromCString` and `toCString` don't care about escapes; escapes are a source-level programming language concern, not a run-time concern. They don't copy byte-by-byte because that would not respect the charset-aware decoding and encoding.